### PR TITLE
Fix #47, use `--persist` to keep SiYuan data files in sandbox and initialize default workspace at first time start, and allow user to read media files and export documents like PDF format.

### LINKED ISSUE #47: Change in build 135223 rejected
A change in [build 135223](https://buildbot.flathub.org/#/builders/6/builds/152137) has been reviewed by the Flathub team (@flathub/build-moderation), and rejected for the following reason:

> The app should use XDG_CONFIG_HOME to access config directories instead of HOME. If it uses HOME please use `--persist` instead of `--filesystem`.
## Changes
| Field | Old value | New value |
| --- | --- | --- |
| filesystems | ['home'] | ['~/.config/SiYuan', '~/.config/SiYuan-Electron', '~/.config/siyuan', '~/SiYuan'] |

### PR DESCRIPTION
**Pull request by Erven2016**
solved: #47

### CODE REVIEW

**Comment by Erven2016** [comment]
> Much appreciated and welcome! Have you tested that? I will test it later.

Yes, I tested and SiYuan works well, you can check a short demonstrated video in issue #47.

But there is a problem is that workspace path was changed, so installed users need to recover their data by moving folder manully.

**Comment by Erven2016** [comment]
You can use (Warehouse)[https://flathub.org/apps/io.github.flattool.Warehouse] or run `rm -rf ~/.var/app/org.b3log.siyuan` to trash user data when testing.

File: start-siyuan.sh [hunk 1]
```diff
@@ -2,6 +2,18 @@
 
 set -oue pipefail
 
+WORKSPACE_CONFIG="/home/$USERNAME/.config/siyuan/workspace.json"
```
[the remark this answers is not in the document]
Comment (Erven2016) [reply start-siyuan.sh:5]: Yes, now I changed this two variable names accroding your suggestions, it looks better than my take.

File: start-siyuan.sh [hunk 2]
```diff
@@ -2,6 +2,18 @@
...
+
+# Creating default workspace configuration when first time opening SiYuan
+# then Siyuan will initialize directly.
+#
+# For users who installed SiYuan, please copy folder `~/SiYuan` in host
+# to `~/.var/app/org.b3log.siyuan/SiYuan` to recover your data, but dont
+# forget to clean up `~/.var/app/org.b3log.siyuan/SiYuan` to empty at first.
+if [ ! -d "$WORKSPACE_CONFIG" ]; then
```
[the remark this answers is not in the document]
Comment (Erven2016) [reply start-siyuan.sh:13]: You are right and sorry for my mistake.

File: org.b3log.siyuan.yml [hunk 3]
```diff
@@ -7,10 +7,20 @@ base-version: *runtime-version
...
+
+  # Allow Siyuan to read media files from host
+  - --filesystem=xdg-pictures:ro
+  - --filesystem=xdg-videos:ro
+  - --filesystem=xdg-music:ro
+  # Allow users export documents to ~/Desktop, ~/Downloads and ~/Documents
+  - --filesystem=xdg-desktop
+  - --filesystem=xdg-download
+  - --filesystem=xdg-documents
+
```
[the remark this answers is not in the document]
Comment (Erven2016) [reply org.b3log.siyuan.yml:23]: Now I removed this blank line, I will push those changes later.

**Comment by Erven2016** [comment]
I added a notification when first time starting SiYuan to notice users where they can export documents.

Screenshot from 2024-10-12 04-06-27

File: org.b3log.siyuan.yml [hunk 4]
```diff
@@ -7,10 +7,18 @@ base-version: *runtime-version
...
+  - --persist=.config/SiYuan
+  - --persist=.config/SiYuan-Electron
+  # Allow Siyuan to read media files from host
+  - --filesystem=xdg-pictures:ro
+  - --filesystem=xdg-videos:ro
+  - --filesystem=xdg-music:ro
+  # Allow users export documents to ~/Desktop, ~/Downloads and ~/Documents
+  - --filesystem=xdg-desktop
+  - --filesystem=xdg-download
+  - --filesystem=xdg-documents
```
Comment (hfiguiere) [inline org.b3log.siyuan.yml:21]:
```suggestion
  - --filesystem=xdg-desktop
  - --filesystem=xdg-download
  - --filesystem=xdg-documents
```

if you can read you can write. no need to be redundant.

File: start-siyuan.sh [hunk 5]
```diff
@@ -2,6 +2,25 @@
...
+WORKSPACE_CONFIG_FILE="/home/$USER/.config/siyuan/workspace.json"
+
+# Creating default workspace configuration when first time opening SiYuan
+# then Siyuan will initialize directly.
+#
+# For users who installed SiYuan, please copy folder `~/SiYuan` in host
+# to `~/.var/app/org.b3log.siyuan/SiYuan` to recover your data, but dont
+# forget to clean up `~/.var/app/org.b3log.siyuan/SiYuan` to empty at first.
+if [ ! -s "$WORKSPACE_CONFIG_FILE" ]; then
+    echo "[\"/home/$USER/SiYuan\"]" > "$WORKSPACE_CONFIG_FILE";
```
Comment (hfiguiere) [inline start-siyuan.sh:14]: this can break anytime if the home dir is not in `/home`. It doesn't have to be.

Use `$HOME`.

Comment (Erven2016) [reply start-siyuan.sh:14]: Okay, thanks your suggestion. I will consider applying it later.

Comment (Erven2016) [reply start-siyuan.sh]: Now I used `$HOME` to locate HomeDir instead of `/home/$USER`.

Thanks for your suggestion again.

**Pull request merged**